packages/hydrooj/src/model/task.ts

```diff
@@ -112,6 +147,10 @@ export async function apply(ctx: Context) {
             expire: new Date(Date.now() + 10000),
         });
     });
+    setInterval(() => {
+        for (const f of waiterQueue) f();
+        waiterQueue.clear();
+    }, 1000);
```

Comment:
still thought the original `Promise.race` would be better as using a waiterQueue there would always be 1s interval between poll

|Timeline| | | | | | | | | | |
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|Promise.race | x | 1 | x | x | 2 | x | 1 | x | x | 2 | 
|setInterval     | x | x | x | x | 1,2 | x | x | x | x | 1,2 |

Reply:
This implementation is the same as the promise race as the end event triggers the notify as well when call back function finishes. So the waiting consumer can be triggered either by end event to set interval as before.

Reply:
No it's different. You are triggering the event for all the waiters at the same time, which would be better to split them into different threads (see timeline above), this can reduce user wait time.

Reply:
Use dynamic waiting time to distribute events evenly.

Reply:
Take the concurrency into consideration to mimic previous behavior when multiple consumer polls the task.